### PR TITLE
feat: add loss in DB model

### PR DESCRIPTION
**Pull request by charlesmindee**
This PR implements DB loss in the compute_loss method.
Any feedback is welcome!

### CODE REVIEW

File: doctr/datasets/loaders/recognition/core.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+class RecognitionDataGenerator(tf.keras.utils.Sequence):
+    """Data loader for recognition model
+
+    Args:
+        input_size: size (h, w) for the images
+        images_path: path to the images folder
+        labels_path: pathe to the json file containing all labels (character sequences)
+        batch_size: batch size to train on
+        suffle: if True, dataset is shuffled between each epoch
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/core.py:46]: typo here "shuffle"

File: doctr/datasets/loaders/recognition/core.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        std_rgb: to normalize dataset
+        std_mean: to normalize dataset
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/core.py:48]: typo here "mean", "std"

File: doctr/datasets/loaders/recognition/core.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    def __init__(
+        self,
+        input_size: Tuple[int, int],
+        images_path: str,
+        labels_path: str,
+        batch_size: int = 64,
+        shuffle: bool = True,
+        std: Tuple[float, float, float] = (0.299, 0.296, 0.301),
+        mean: Tuple[float, float, float] = (0.694, 0.695, 0.693),
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/core.py:59]: mind inversing the order please? (usually mean, then std in most frameworks/libraries)

File: doctr/datasets/loaders/recognition/core.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            np.random.shuffle(self.indexes)
+
+    def __getitem__(
+        self,
+        index: int
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        # Get one batch of data
+        indexes = self.indexes[index * self.batch_size:min(self.__len__(), (index + 1) * self.batch_size)]
+        # Find list of paths
+        list_paths = [os.listdir(self.images_path)[k] for k in indexes]
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/core.py:87]: We probably should build `os.listdir(self.images_path)` in the constructor
That way, we could also update all other references to it

File: test/test_models_detection.py [hunk 5]
```diff
@@ -180,3 +181,14 @@ def test_compute_target_db():
...
+    return gts, masks, thresh_gts, thresh_masks
+
+
+def test_compute_loss_db(test_compute_target_db):
+    model = detection.db_resnet50(pretrained=False)
+    dbinput = tf.random.uniform(shape=[3, 1024, 1024, 3], minval=0, maxval=1)
+    proba_map, thresh_map, binary_map = model(dbinput, training=True)
+    gt, mask, thresh_gt, thresh_mask = test_compute_target_db
+    loss = model.compute_loss(proba_map, binary_map, thresh_map, gt, mask, thresh_gt, thresh_mask)
+    assert isinstance(loss, tf.Tensor)
```
Comment (fg-mindee) [inline test/test_models_detection.py:194]: Perhaps we should also check that this is a valid loss (i.e. that we can backprop it?)

File: doctr/datasets/loaders/recognition/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+from .core import *
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/__init__.py:1]: You can delete this file

File: doctr/datasets/loaders/recognition/core.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# Copyright (C) 2021, Mindee.
```
Comment (fg-mindee) [inline doctr/datasets/loaders/recognition/core.py:1]: Let's move the file to `loaders/recognition.py`

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: test/test_models_detection.py [hunk 8]
```diff
@@ -187,8 +187,12 @@ def test_compute_target_db():
...
     gt, mask, thresh_gt, thresh_mask = test_compute_target_db
-    loss = model.compute_loss(proba_map, binary_map, thresh_map, gt, mask, thresh_gt, thresh_mask)
-    assert isinstance(loss, tf.Tensor)
+    optimizer = tf.keras.optimizers.SGD(learning_rate=1e-3)
+    with tf.GradientTape() as tape:
+        proba_map, thresh_map, binary_map = model(db_input, training=True)
+        loss = model.compute_loss(proba_map, binary_map, thresh_map, gt, mask, thresh_gt, thresh_mask)
+        assert isinstance(loss, tf.Tensor)
+    grads = tape.gradient(loss, model.trainable_weights)
+    optimizer.apply_gradients(zip(grads, model.trainable_weights))
```
Comment (fg-mindee) [inline test/test_models_detection.py:198]: Perhaps we could try to remove only the update? (if the backprop only is working, that's enough)
If not perhaps freezing the backbone would help! (backprop only through the last layers)

**Review by fg-mindee — COMMENTED** [review]
I added some suggestions to fix the CI jobs

**Comment by charlesmindee** [comment]
Unfortunately we can't add backpropagation in CI without crashing it

**Review by fg-mindee — APPROVED** [review]
Thanks for trying to check the backprop, looks good to me as is!
